Adding CPU training support to AxoNN

Avuxon opened the pull request:
- Switch to gloo for communication

Avuxon commented:
Need to make the changes conditional on CPUs/GPUs

siddharth9820 commented on axonn/communication.py:
```diff
@@ -45,8 +46,7 @@ def __init__(
         )
-        self.local_rank = self.world_rank % self.gpus_per_node
-        torch.cuda.set_device(self.local_rank)
+        self.local_rank = os.environ["SLURM_LOCALID"]
```
Are we using the local rank anywhere? We can just get rid of this line. Local Rank was required in the GPU case to pick the right GPU.

Dando18 commented:
Before merging this should

- [ ] modular cpu support; the existing code was okay for gpu setups; the library should take a flag `cpu_only` to run in cpu mode; or it should auto-detect what it should do
- [ ] add tests; this should be able to run on github runners since it's cpu

Dando18 commented on axonn/communication.py:
```diff
@@ -31,6 +31,17 @@ def __init__(
+        if gpus_per_node is None:
+            self.backend = 'gloo'
+            self.is_gpu_available = False
+            self.local_rank = os.environ.get("SLURM_LOCALID", 0)
```
I don't think `local_rank` gets used if it's on cpu, so we can probably get rid of this.

siddharth9820 replied:
yeah this can go away

Dando18 commented on axonn/communication.py:
```diff
@@ -108,15 +113,15 @@ def __init__(
                     group_members = intra_layer_ranks[i::G_intra_c]
                     group = torch.distributed.new_group(
-                        ranks=group_members, backend="nccl"
+                        ranks=group_members, backend="gloo"
```
Should this be `self.backend`?

siddharth9820 replied:
yep

siddharth9820 replied:
backend initiailization - 
if torch.cuda.is_available() - nccl
elif torch.distributed.is_mpi_available() - mpi
else gloo

Commit:
Rebased cpu changes onto latest axonn

Commit:
Args should default to true, removed tensor_list operator for cpus

Commit:
Made changes conditional for CPUs/GPUs

Commit:
Gpu runs do not make Slurm calls

S-Mahua committed:
removed mpi4py dependency (#63)

Co-authored-by: Mahua Singh <mahua04@pssg-mordor.umiacs.umd.edu>

Commit:
Revamped cpu flag with cpu hard set in comm.py

Add default values for environment vars

Fixed communication handle flags

Fixed formatting

jwendlan committed:
adding parallelize context for opt (#65)

siddharth9820 committed:
Removing the drop and gathers in depth tensor parallelism for the easy API (#66)

siddharth9820 committed:
change parallelize context to use AutoConfig (#67)

Commit:
Set device before init

Formatting fixed

formatting again

tensor-list change

Formatting and Device-Setting

Fixed gpus_per_node access

adityaranjan committed:
Bugfix: Initialize grad_input, grad_weight to None (#68)

* initialize grad_input to None

* minor

Commit:
Added initial pytest parameters for cpu tests

Commit:
Removed distributed communication class

Added missing parameter to test

Fixed formatting

docs: fix build issues and add sub-sections (#69)

siddharth9820 committed:
Merge branch 'develop' into axonn-cpu

siddharth9820 committed:
formatting

siddharth9820 committed:
fix CI

siddharth9820 committed:
format

siddharth9820 reviewed (commented):
When attempting to fix the CI I discovered that the gloo backend doesn't support reduce-scatters. Therefore, as of now AxoNN wouldn't work on CPUs with G_intra_d>1. We should add an assert that checks this condition in axonn.py.

siddharth9820 commented on .github/workflows/nvidia-rtx-3090-tests.yaml:
```diff
@@ -17,6 +17,10 @@ jobs:
         memopt: [ '0', '1' ]
 
+    env:
+      SLURM_NTASKS: 0
```
let's remove these SLURM variables

Commit:
Matched changes for intra_layer_conv tests

Commit:
Removed unused env variables and format

Avuxon committed:
Merge branch 'develop' into axonn-cpu

Commit:
Fixed CI to return for depth > 1

Commit:
Merge branch 'axonn-cpu' of https://github.com/axonn-ai/axonn into axonn-cpu

Commit:
Added depth check to fw pass in fc test

Commit:
Merge develop into this branch

Commit:
Added xfail to convolution tests

siddharth9820 committed:
remove xfail

siddharth9820 committed:
Merge branch 'develop' into axonn-cpu

siddharth9820 committed:
Merge branch 'develop' into axonn-cpu
